[rpc] special case tensor type check when getting RRef

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33263 [jit] allow RRef local creation with IValue objects
* **#33582 [rpc] special case tensor type check when getting RRef**
* #33526 [jit] make RRef type annotation available in Python
* #33369 [jit] infer RRef type as container type

Differential Revision: [D20009837](https://our.internmc.facebook.com/intern/diff/D20009837)

wanchaol committed:
[rpc] special case tensor type check when getting RRef

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] special case tensor type check when getting RRef"


Differential Revision: [D20009837](https://our.internmc.facebook.com/intern/diff/D20009837)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/rref_context.cpp:
```diff
@@ -175,8 +175,22 @@ c10::intrusive_ptr<RRef> RRefContext::getOrCreateRRef(
+    }
     return ownerRRef;
   } else {
     return createUserRRef(ownerId, rrefId, forkId, type);
```
how about createUserRRef case when type is not exactly matched

wanchaol replied:
iirc `createUserRRef` will not try to find if there's an existing UserRRef in the RRefContext, so there would be no type match problem.

zhaojuanmao replied:
I mean whether there will be an issue if one user created user rref with subtypeOf(TensorType::get()), then the user shared this user rref to another user, and another user will create user rref here with a TensorType::get(). iiuc, the owner rref will have subtypeof(TensorType::get()). 

So some user rref will have slightly different type based on above, I'm wondering whether this will be an issue?

wanchaol replied:
The plain TensorType is always a subtype of specialized SubTensorType, the only reason why this is failing is that we have a Type equal assertion here,  I didn't see this in other places. For the case that you described, the `forked UserRRef` will holding the plain TensorType, which is subtype compatible with the SubTensorType, which should be safe. In fact, we can only get into your described case when we fist run the ScriptFunction locally, then call `rpc.remote` on this ScriptFunction again remotely. But when we run the ScriptFunction in remote, we shouldn't preserve the Specialized SubTensorType information (because that's the information we get from the local run). So I think the fix here should be enough.

zhaojuanmao replied:
thanks for explaining! Sounds good, would you please also add this to comment?

zhaojuanmao reviewed (commented):
there are real test failures, please see inline comments as well

wanchaol committed:
Update on "[rpc] special case tensor type check when getting RRef"


Differential Revision: [D20009837](https://our.internmc.facebook.com/intern/diff/D20009837)

[ghstack-poisoned]

wanchaol committed:
Update on "[rpc] special case tensor type check when getting RRef"


Differential Revision: [D20009837](https://our.internmc.facebook.com/intern/diff/D20009837)

[ghstack-poisoned]
